pkg/admission/karydia/karydia.go

```diff
@@ -48,129 +46,146 @@ func New(config *Config) (*KarydiaAdmission, error) {
 	if err != nil {
-		e := fmt.Errorf("failed to determine pod's namespace: %v", err)
-		k.logger.Errorf("%v", e)
-		return k8sutil.ErrToAdmissionResponse(e)
+		k.logger.Errorf("%v", err)
+		return k8sutil.ErrToAdmissionResponse(err)
 	}
 
-	automountServiceAccountToken, doCheck := namespace.ObjectMeta.Annotations["karydia.gardener.cloud/automountServiceAccountToken"]
-	if doCheck {
-		if automountServiceAccountToken == "forbidden" {
-			if doesAutomountServiceAccountToken(&pod) {
-				validationErrors = append(validationErrors, "automount of service account not allowed")
-			}
-		} else if automountServiceAccountToken == "non-default" {
-			if doesAutomountServiceAccountToken(&pod) && pod.Spec.ServiceAccountName == "default" {
-				validationErrors = append(validationErrors, "automount of service account 'default' not allowed")
-			}
-		}
+	automountServiceAccountToken, annotated := namespace.ObjectMeta.Annotations["karydia.gardener.cloud/automountServiceAccountToken"]
+	if annotated {
+		patches, validationErrors = admitServiceAccountToken(*pod, automountServiceAccountToken, patches, validationErrors)
 	}
 
-	seccompProfile, doCheck := namespace.ObjectMeta.Annotations["karydia.gardener.cloud/seccompProfile"]
-	if doCheck {
-		seccompPod, ok := pod.ObjectMeta.Annotations["seccomp.security.alpha.kubernetes.io/pod"]
-		if !ok && mutationAllowed {
-			if len(pod.ObjectMeta.Annotations) == 0 {
-				// If no annotation object exists yet, we have
-				// to create it. Otherwise we will encounter
-				// the following error:
-				// 'jsonpatch add operation does not apply: doc is missing path: "/metadata/annotations..."'
-				patches = append(patches, fmt.Sprintf(`{"op": "add", "path": "/metadata/annotations", "value": {"%s": "%s"}}`, "seccomp.security.alpha.kubernetes.io/pod", seccompProfile))
-			} else {
-				patches = append(patches, fmt.Sprintf(`{"op": "add", "path": "/metadata/annotations/%s", "value": "%s"}`, "seccomp.security.alpha.kubernetes.io~1pod", seccompProfile))
-			}
-		} else if seccompPod != seccompProfile {
-			validationErrorMsg := fmt.Sprintf("seccomp profile ('seccomp.security.alpha.kubernetes.io/pod') must be '%s'", seccompProfile)
-			validationErrors = append(validationErrors, validationErrorMsg)
-		}
+	seccompProfile, annotated := namespace.ObjectMeta.Annotations["karydia.gardener.cloud/seccompProfile"]
+	if annotated {
+		patches, validationErrors = admitSeccompProfile(*pod, seccompProfile, mutationAllowed, patches, validationErrors)
 	}
 
-	if len(validationErrors) > 0 {
-		return &v1beta1.AdmissionResponse{
-			Allowed: false,
-			Result: &metav1.Status{
-				Message: fmt.Sprintf("%+v", validationErrors),
-			},
+	return admitResponse(patches, validationErrors)
+}
+
+func admitServiceAccountToken(pod corev1.Pod, annotation string, patches []string, validationErrors []string) ([]string, []string) {
+	if annotation == "forbidden" {
+		// Validating webhook
+		if automountServiceAccountTokenUndefined(&pod) {
```

Comment:
If I set `AutomountServiceAccountToken` to `true`, the service account token will be mounted, even with

```
karydia.gardener.cloud/automountServiceAccountToken: forbidden
```

set on the namespace.

Probably due to the change to `doesAutomountServiceAccountToken` below.

Reply:
@Neumann-Nils neels pointed that out in https://github.com/karydia/karydia/pull/66#pullrequestreview-223232023 and instead of fixing the bug the test was changed?

Reply:
If not explicitly set, Kubernetes sets the `AutomountServiceAccountToker` to `true`. Since we want to provide a secure-by-default behavior we changed that. But if a developer explicitly sets this value to `true`, you can assume that he knows what he's doing and therefore automounting is allowed.

Reply:
@marwinski What do you think about these two options?

Reply:
> If not explicitly set, Kubernetes sets the AutomountServiceAccountToker to true.

By default, it's unset (i.e. `nil` in Go land) but the token mounted. For example

```
# Terminal 1; note the service account token at /var/run/secrets/kubernetes.io/serviceaccount/token
kubectl run --rm -ti --restart=Never --image busybox busybox

# Terminal 2
kubectl get pod busybox -o jsonpath='{.spec.automountServiceAccountToken}'
```

No output here, would be `true` if `automountServiceAccountToken` is actually set.

What karydia currently does is removing the token volume from the pod after the fact: https://github.com/karydia/karydia/blob/ad19f38a30b79663ae71242ff466e55d930773ba/pkg/admission/karydia/karydia.go#L112-L123

A previous discussion can be found here: https://github.com/karydia/karydia/issues/55

> But if a developer explicitly sets this value to true, you can assume that he knows what he's doing and therefore automounting is allowed.

Hmm, I'm skeptical about that: the configuration value says `forbidden`. If an administrator installs karydia and sets the configuration to `forbidden`, their intention is very clear: "do not allow auto mounting the service account token". Making it a `maybe forbidden` IMHO is confusing and counter intuitive.

I'm not sure a new value `remove-default` is actually needed (and if that could be concluded from the other settings instead). Maybe a matrix that shows all possible constellations would help to find an answer here.

In pseudo-code, I wonder if the following would work:

```
if automountServiceAccountToken == nil || *automountServiceAccountToken == true
  if forbidden
    return error
  else if nonDefault && name == "default"
    if mutating
      return patchRemoveDefault
    else
      return error
```

i.e. `forbidden` forbids auto mounting of any service account token and `non-default` forbids auto mounting of the default service account token.

Reply:
> If an administrator installs karydia and sets the configuration to `forbidden`, their intention is very clear: "do not allow auto mounting the service account token".

You've definitely got a point here. This may cause confusion for an administrator. Probably we should find another name for the secure-by-default behavior and stay with `forbidden` for the prohibitive setting.

> I'm not sure a new value `remove-default` is actually needed (and if that could be concluded from the other settings instead).

From a behavioral perspective, this is actually requested. Code-wise, we would like to have a clear separation of validation and mutating webhook. Still, the behavior of a setting should not depend on which webhook is registered as it would be in for the pseudo-code example. This would also be quite confusing for an admin. 

We propose to have both, mutating and validating webhooks registered without exception. Depending on the annotation they mutate and validate or only validate. In fact, the secure-by-default options should mutate and validate and the prohibitive options should probably only validate. 

Options may look like the follows:

- `forbidden`  *validating* no service accounts token mounts allowed
- `non-default`   *validating* no default service accounts token mounts allowed
- `remove-all` *mutating & validating* no service account tokens when not explicitly enabled
- `remove-default` *mutating & validating* no default service account token when not explicitly enabled

Just as a proposition. Ideas & comments on this @schu ?

Reply:
Looping in @CodeClinch. Any ideas/comments on the proposed options?